pkg/segment/writer/metrics/wal/wal.go

```diff
@@ -0,0 +1,272 @@
+	if errors.Is(err, io.EOF) {
+		return nil, false, nil
+	} else if err != nil {
+		log.Errorf("Next: failed to read block size from WAL file: %v", it.fd.Name())
+		return nil, false, err
+	}
+
+	it.readBuf = toputils.ResizeSlice(it.readBuf, int(blockSize))
+	_, err = io.ReadFull(it.fd, it.readBuf)
+	if err != nil {
+		log.Errorf("Next: failed to read block data of size %d from file %s: %v", blockSize, it.fd.Name(), err)
+		return nil, false, err
+	}
+
+	blockBuf := bytes.NewReader(it.readBuf)
+	newBlock, err := it.decodeWALBlock(blockBuf)
+
+	if err != nil {
+		log.Errorf("Next: dataDecompression failed: %v", err)
+		return nil, false, err
+	}
+
+	it.currentBlock = newBlock
+	it.currentIndex = 1
+
+	return &newBlock[0], true, nil
+}
+
+func (it *WalIterator) Close() error {
+	return it.fd.Close()
+}
+
+func (w *WAL) encodeWALBlock(dps []walDatapoint, compressedBlockBuf *bytes.Buffer) error {
+	N := uint32(len(dps))
+	if N == 0 {
+		log.Warn("EncodeWALBlock: received empty data points slice")
+		return errors.New("empty data points")
+	}
+
+	rawBlockBuf := &bytes.Buffer{}
+
+	err := binary.Write(rawBlockBuf, binary.LittleEndian, N)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write datapoint count: %v", err)
+		return err
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(rawBlockBuf, binary.LittleEndian, dp.timestamp); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write timestamp: %v", err)
+			return err
+		}
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(rawBlockBuf, binary.LittleEndian, dp.dpVal); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write dpVal: %v", err)
+			return err
+		}
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(rawBlockBuf, binary.LittleEndian, dp.tsid); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write tsid: %v", err)
+			return err
+		}
+	}
+
+	w.encodeBuf = encoder.EncodeAll(rawBlockBuf.Bytes(), w.encodeBuf[:0])
+
+	_, err = compressedBlockBuf.Write(w.encodeBuf)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write compressed data to buffer: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (it *WalIterator) decodeWALBlock(blockBuf *bytes.Reader) ([]walDatapoint, error) {
+	compressedData, err := io.ReadAll(blockBuf)
+	if err != nil {
+		log.Errorf("decodeWALBlock: failed to read compressed data: %v", err)
+		return nil, err
+	}
+
+	it.readBuf, err = decoder.DecodeAll(compressedData, it.readBuf[:0])
+
+	if err != nil {
+		log.Errorf("decodeWALBlock: decompression failed: %v", err)
+		return nil, err
+	}
+
+	rawReader := bytes.NewReader(it.readBuf)
+
+	var N uint32
+	err = binary.Read(rawReader, binary.LittleEndian, &N)
+	if err != nil {
+		log.Errorf("decodeWALBlock: failed to read datapoint count: %v", err)
+		return nil, err
+	}
+
+	timestamps := make([]uint64, N)
+	dpVals := make([]float64, N)
+	tsids := make([]uint64, N)
+
+	for i := 0; i < int(N); i++ {
+		err := binary.Read(rawReader, binary.LittleEndian, &timestamps[i])
```

Comment:
then in here you would do :
` binary.Read(rawReader, binary.LittleEndian, &it.readDps[i].timestamp)`